src/db.c

```diff
@@ -56,14 +57,46 @@ void updateLFU(robj *val) {
     val->lru = (LFUGetTimeInMinutes()<<8) | counter;
 }
 
-/* Low level key lookup API, not actually called directly from commands
- * implementations that should instead rely on lookupKeyRead(),
- * lookupKeyWrite() and lookupKeyReadWithFlags(). */
+/* Lookup a key for read or write operations, or return NULL if the key is not
+ * found in the specified DB. This function implements the functionality of
+ * lookupKeyRead(), lookupKeyWrite() and their ...WithFlags() variants.
+ *
+ * Side-effects of calling this function:
+ *
+ * 1. A key gets expired if it reached it's TTL.
+ * 2. The key's last access time is updated.
+ * 3. The global keys hits/misses stats are updated (reported in INFO).
+ * 4. If keyspace notifications are enabled, a "keymiss" notification is fired.
+ *
+ * Flags change the behavior of this command:
+ *
+ *  LOOKUP_NONE (or zero): No special flags are passed.
+ *  LOOKUP_NOTOUCH: Don't alter the last access time of the key.
+ *  LOOKUP_NONOTIFY: Don't trigger keyspace event on key miss.
+ *  LOOKUP_NOSTATS: Don't increment key hits/misses couters.
+ *  LOOKUP_NOEXPIRE: Don't check if the key has expired.
+ *  LOOKUP_WRITE: Prepare the key for writing (delete expired keys even on
```

Comment:
one of the purposes of this PR is to solve problems in writable replicas, they're an odd configuration but i think for now we need to keep supporting them (unlike slaveKeysWithExpire which i care even less for).

i.e. the other purpose of this PR is a cleanup for code that makes more sense.

note that LOOKUP_WRITE may have other purposes soon, like updating write hit / miss statistics maybe.

I don't recall why we use LOOKUP_NOEXPIRE in SWAPDB, maybe it's just to prevent unintended changes of the refactoring? @zuiderkwast do you recall?
but anyway, i can think of one reason not to expire in this case: it can cause a flood of expiration that will freeze the server and cause replicas to drop.

Reply:
That's right, NOEXPIRE is just to keep the exact same behaviour as before when refactoring. (It's used in a function that calls lookupKey directly, which didn't do expire before. Expire was done in lookupKeyRead and lookupKeyWrite only.)

I'm trying now to delete the NOEXPIRE flag and run the tests locally. It seems to pass, but I'll run the cluster and module tests too. If it works, we can get rid of this flag.

Reply:
I have deleted NOEXPIRE.

The WRITE flag is for writable replicas, it's true. I also don't like them, but if we can't deprecate them, it's better to make them a little more consistent. Also, maybe you can accept oran's explanation:

> note that LOOKUP_WRITE may have other purposes soon, like updating write hit / miss statistics maybe.

Reply:
i have some concern about this change, i'm not certain this behavior of not doing expiration on SWAPDB wasn't on purpose.
In theory, if we would have scanned a lot of keys, we can cause a freeze to the server and blow up the replication buffers with DELs.. expiration efforts must be gradual and not all in one go.
In practice we only scan keys that some clients are being blocked on, so maybe it's not that many keys.
also possibly when we return from the SWAPDB command, we'll lookup these keys again and expire them anyway, is that right?
@zuiderkwast can you look into that, and maybe also into the blame log and history of SWAPDB and scanDatabaseForReadyLists?

Reply:
Yes, a useless lookup of the expire dict is the potentially costly part here.

We can preserve the old behaviour here if we just lookup the dict manually, without calling lookupKey. With all those flags, the only thing we actually want to do here is look up the dict anyway. I can do that.

Btw, my plan is to update the 'hasexpire' PR when this one is merged.

Reply:
> Yes, a useless lookup of the expire dict is the potentially costly part here.

you mean that if the key already expired, then this event loop cycle will release it (possibly slow operation for a complex key), and propagate DEL anyway.. i.e. it'll happen in scanDatabaseForReadyLists instead of handleClientsBlockedOnKeys, right?

let's add a test for that if there isn't already one.
i.e. make sure we have a test for `SWAPDB that awakes a blocked client`, and also write one for `SWAPDB that awakes blocked client, but the key already expired`.

> We can preserve the old behaviour here if we just lookup the dict manually, without calling lookupKey. With all those flags, the only thing we actually want to do here is look up the dict anyway. I can do that.

i suppose that's a good option if we anyway try to avoid all the checks and side-effects of lookupKey.
i.e. Technically this function isn't done on behalf of a command, it's not really looking up the key but rather checks what's in the dict. so i vote for it.

> Btw, my plan is to update the 'hasexpire' PR when this one is merged.

sure, that's the next step (it's the one that has awaken this PR)

Reply:
SWAPDB tests added. Using dictFind instead of lookupKey in scanDatabaseForReadyLists. I think it's ready to merge unless there is anything else.